rsksmart/rwallet
# feat: locks passcode screen on too many wrong consecutive attempts

## Features
- Avoids users to enter passcodes on too many consecutive wrong attempts
- Adds countdown timer to show when users can retry
- Handles steps for retries dynamically
- Persists wrong attempts count and last attempt timestamp between sessions
- Adds loader while checking if user can already retry

## Known issues
- I just got the translation from Google so not sure if they are the most accurate
- It might be better to map props to state and read them on `componentWillReceiveProps` instead of using `baseModal.resetModal(...)`, but that would require a refactor of the base modal dependencies as well (reset passcode, create new passcode, wrong passcode

## How to test
- Wrong retries count and timer should persist between sessions (ie: closing and re-opening the app)
- With the provided configuration, the app should behave like this:
```
    STEP_1
    6 consecutive -> 1 minute wait
An iPhone will disable for 1 minute after six failed passcode attempts in a row
STEP_2
    7 consecutive -> 5 minutes
STEP_3
    8 consecutive -> 15 minutes
STEP_4
    10 or more consecutive -> 1 hour
```
- The steps can be changed on `src/components/common/passcode/wrongPasscodeUtils.js` -> `WRONG_ATTEMPTS_STEPS`

## Screenshots
<img width="346" alt="Screen Shot 2021-04-22 at 17 23 54" src="https://user-images.githubusercontent.com/8449567/115780470-94617d80-a38f-11eb-811f-2b26fb90db3d.png">
